### PR TITLE
ocs-operator: Nightly images

### PR DESCRIPTION
**Pull request by kshlm**
This PR creates a nightly job that builds ocs-operator images.

Next steps will be to add push secret for quay.io/ocs-dev to the openshift-ci
secret store and add ocs-operator to the image mirroring config.

### CODE REVIEW

**Comment by kshlm** [comment]
The release-4.2 jobs are failing, as they haven't been updated in about 3 weeks, and do not contain the CI changes added by @davidvossel. We have two options, either I remove the changes done here for the 4.2 branch, or we push the new changes in ocs-operator:master to ocs-operator:release-4.2.

The release-4.3 and release-4.4 branches currently track master, so they already have the required changes.

**Comment by kshlm** [comment]
Anyways, the failed jobs are not `required checks`, so we can still merge this PR.

**Comment by davidvossel** [comment]
> We have two options, either I remove the changes done here for the 4.2 branch, or we push the new changes in ocs-operator:master to ocs-operator:release-4.2

is there any reason why we shouldn't sync the 4.2 branch now?  I thought 4.2 would ultimately be the stable branch we start porting patches into after dev freeze

**Comment by jarrpa** [comment]
> > We have two options, either I remove the changes done here for the 4.2 branch, or we push the new changes in ocs-operator:master to ocs-operator:release-4.2
> 
> is there any reason why we shouldn't sync the 4.2 branch now? I thought 4.2 would ultimately be the stable branch we start porting patches into after dev freeze

I'm also confused, why is release-4.2 not tracking master as well? We have no reason to be diverging yet.

**Comment by jarrpa** [comment]
The PR overall looks good. I'd like to get a little more understanding about the release-4.2 end of things before giving the LGTM.

**Comment by kshlm** [comment]
> All of the ocs-registry container images, whether they are built by master, release 4.2, 4.3, 4.4, will point to an ocs-operator image in quay.io/ocs-dev/ocs-operator:latest.
> 
> Is my understanding here correct? Do we need to have a way to separate latest for each branch?

When we set up image mirroring, we'll have images with the following floating tags - latest, 4.2.0, 4.3.0, 4.4.0 - tracking respectively, master, release-4.2, release-4.3 and release-4.4 branches. As master, release-4.3 and release-4.4 are the same branch now, the registry images based off them will all point to ocs-operator:latest.
For release-4.2, which has diverged from master now, we will have to fix the registry image and have it point to ocs-operator:4.2.0. I'll send a PR to change the OCS_IMAGE used, after we rebase release-4.2 to master. The same will need to be done when release-4.{3,4,..} diverge from master.

**Comment by kshlm** [comment]
> I'm also confused, why is release-4.2 not tracking master as well? We have no reason to be diverging yet.

This is most likely how the Openshift developer platform team have set up repositories under the Openshift organization. As I see they always have the next two release branches (+1 and +2) tracking master. When the dev freeze for the next release happens,  the +1 branch stop tracking master and a new +3 branch is created tracking master.

When the 4.2 OCP freeze happened, the ocs-operator repo also had its release-4.2 branch detached from master and a new release-4.4 branch created tracking master. As we've only now reached dev freeze for OCS, we'll have to manually rebase release-4.2 onto current master.

**Comment by davidvossel** [comment]
> When we set up image mirroring, we'll have images with the following floating tags - latest, 4.2.0, 4.3.0, 4.4.0 - tracking respectively, master, release-4.2, release-4.3 and release-4.4 branches. As master, release-4.3 and release-4.4 are the same branch now, the registry images based off them will all point to ocs-operator:latest.

yeah... but how is the CSV changed to point to the right ocs-operator image. we'll tag ocs-registry to 4.2.0 tag for example, but the CSV still points to latest. right?

**Comment by davidvossel** [comment]
> yeah... but how is the CSV changed to point to the right ocs-operator image. we'll tag ocs-registry to 4.2.0 tag for example, but the CSV still points to latest. right?

I literally just repeated your last sentence. Sorry

How do we plan to fix that the CSV points to latest for 4.2, 4.3, 4.4?  are we going to make a change to the CSV just in those branches? that would be fine with me. we just have to remember to make that change each time a 4.x branch is cut.

**Comment by kshlm** [comment]
> yeah... but how is the CSV changed to point to the right ocs-operator image. we'll tag ocs-registry to 4.2.0 tag for example, but the CSV still points to latest. right?

In the ocs-operator repo, `OCS_IMAGE` is set to `ocs-operator:latest` in the `gen-latest-csv.sh` script. In the release-4.2 branch, we'll have to change this to `ocs-operator:4.2.0`.

**Comment by davidvossel** [comment]
> In the ocs-operator repo, OCS_IMAGE is set to ocs-operator:latest in the gen-latest-csv.sh script. In the release-4.2 branch, we'll have to change this to ocs-operator:4.2.0

got it

/lgtm

File: ci-operator/config/openshift/ocs-operator/openshift-ocs-operator-master.yaml [hunk 1]
```diff
@@ -20,38 +27,88 @@ images:
...
+        source_path: /go/src/github.com/openshift/ocs-operator/must-gather/templates
+  to: ocs-must-gather-nightly
+  optional: true
+
 promotion:
   additional_images:
     ocs-operator-tests: src
+  excluded_images:
+  - ocs-operator
+  - operator-registry-base
```
Comment (jarrpa) [inline ci-operator/config/openshift/ocs-operator/openshift-ocs-operator-master.yaml]: For `excluded_images`, some files are using `operator-registry-base` and some are using `ocs-registry-base`. Is this intentional? The former looks like it'd be more correct.

Comment (kshlm) [reply ci-operator/config/openshift/ocs-operator/openshift-ocs-operator-master.yaml]: You're right, I forgot to update the configs for the other branches after I fixed the master config.

Comment (kshlm) [reply ci-operator/config/openshift/ocs-operator/openshift-ocs-operator-master.yaml]: Fixed in #5407

**Pull request merged**